pep-0646.rst

```diff
@@ -733,28 +733,159 @@ hopefully enable a thriving ecosystem of tools for analysing and verifying
-=======================
+Grammar Changes
+===============
+
+This PEP would require two grammar changes. Full diffs of ``python.gram``
+and simple tests to confirm correct behaviour are available at
+https://github.com/mrahtz/cpython/commits/pep646-grammar.
+
+Star expressions in indexes
+---------------------------
+
+The first grammar change enables use of star expressions in index operations (that is,
+within square brackets), necessary to support star-unpacking of TypeVarTuples.
+
+Before:
+
+::
+    
+    slices[expr_ty]:
+        | a=slice !',' { a }
+        | a[asdl_expr_seq*]=','.slice+ [','] { _PyAST_Tuple(a, Load, EXTRA) }
+
+After:
+
+::
+    
+    slices[expr_ty]:
+        | a=slice !',' { a }
+        | a[asdl_expr_seq*]=','.(slice | starred_expression)+ [','] { _PyAST_Tuple(a, Load, EXTRA) }
+
+As with star-unpacking in other contexts, the star operator would call ``__iter__`` on the callee,
+and add the contents of the resulting iterator to the list of index arguments.
+This would enable ``TypeVarTuple`` to be implemented as roughly:
+
+::
 
-In order to use the star operator for unpacking of ``TypeVarTuple`` instances,
-we would need to make two grammar changes:
+    class TypeVarTuple:
+        def __iter__(self):
+            yield _UnpackedTypeVarTuple()
 
-1. Star expressions must be made valid in at least index operations.
-   For example, ``Tuple[*Ts]`` and ``Tuple[T1, *Ts, T2]`` would both
-   be valid. (This PEP does not allow multiple unpacked ``TypeVarTuple``
-   instances to appear in a single parameter list, so ``Tuple[*Ts1, *Ts2]``
-   would be a runtime error. Also note that star expressions would *not*
-   be valid in slice expressions -  e.g. ``Tuple[*Ts:*Ts]`` is
-   nonsensical and should remain invalid.)
-2. We would need to make '``*args: *Ts``' valid in function definitions.
+There are three things to note about this grammar change. First, for consistency with
+star-unpacking in other contexts, ``_UnpackedTypeVarTuple`` would always be received
+by ``__class_getitem__`` as part of a tuple:
+
+::
+
+    class C:
+        def __class_getitem__(self, x):
+            ...
+            
+    [*(1,)]      # = [1]
+    [*(1,), 2]   # = [1, 2]
+    C[int]       # __class_getitem__ receives x=int
+    # But perhaps surprisingly:
+    C[*Ts]       # __class_getitem__ receives x=(_UnpackedTypeVarTuple, )
+    # This is consistent with:
+    C[*Ts, int]  # __class_getitem__ receives x=(_UnpackedTypeVarTuple, int)
+
+
+Second, the change would also enable star-unpacking of other structures within
+indexing operations:
+
+::
+
+    idxs_to_select = (1, 2, 3)
+    array[0, *idxs_to_select, -1]
+
+Third, more than one instance of a star-unpack can occur within an index:
+
+::
+
+    array[*idxs_to_select, *idxs_to_select]
+
+The requirement that only one unpacked TypeVarTuple may be used per list of type parameters
+would therefore be implemented in e.g. ``typing.py`` rather than at the syntax level.
```

Comment:
I'd suggest it should be implemented only by static type checkers, not necessarily by typing.py.

Reply:
Right; in the future, when type annotations are not evaluation at runtime at all (maybe this is the case for 3.11 already, I've lost track of when this was supposed to happen?), this will only be validation by static type checkers. For the time being, though, when type annotations _are_ still being evaluated at runtime, it seems more consistent with runtime-validation of other typing-related stuff to also implement some checking in typing.py.

Reply:
It's still up in the air what will happen in 3.11; we may yet go with https://www.python.org/dev/peps/pep-0649/ .

Regardless, there are many invalid annotations that typing.py doesn't reject, and my general experience is that such checks tend to cause problems with future enhancements to the type system.

Reply:
Fair enough; I've amended this paragraph to state this should be checked by type checkers, and removed reference to typing.py.